magik-checks/src/main/java/nl/ramsolutions/sw/magik/checks/checks/NestingDepthCheck.java

```diff
@@ -0,0 +1,69 @@
+package nl.ramsolutions.sw.magik.checks.checks;
+
+import com.sonar.sslr.api.AstNode;
+import java.util.ArrayDeque;
+import java.util.Deque;
+import nl.ramsolutions.sw.magik.checks.MagikCheck;
+import org.sonar.check.Rule;
+import org.sonar.check.RuleProperty;
+
+/** Check nesting depth of nodes. */
+@Rule(key = NestingDepthCheck.CHECK_KEY)
+public class NestingDepthCheck extends MagikCheck {
+
+  @SuppressWarnings("checkstyle:JavadocVariable")
+  public static final String CHECK_KEY = "NestingDepth";
+
+  private static final int DEFAULT_MAXIMUM_NESTING_DEPTH = 3;
+  private static final String MESSAGE = "The nesting depth is greater than permitted (%s).";
+
+  /** Maximum nesting depth of node. */
+  @RuleProperty(
+      key = "maximum nesting depth",
+      defaultValue = "" + DEFAULT_MAXIMUM_NESTING_DEPTH,
+      description = "Maximum nesting depth",
+      type = "INTEGER")
+  @SuppressWarnings("checkstyle:VisibilityModifier")
+  public int maximumNestingDepth = DEFAULT_MAXIMUM_NESTING_DEPTH;
+
+  private Deque<AstNode> depthNodes = new ArrayDeque<>();
+
+  @Override
+  protected void walkPreIf(final AstNode node) {
+    this.checkNestingDepth(node);
+  }
+
+  @Override
+  protected void walkPostIf(final AstNode node) {
+    depthNodes.pop();
+  }
+
+  @Override
+  protected void walkPreLoop(final AstNode node) {
+    this.checkNestingDepth(node);
+  }
+
+  @Override
+  protected void walkPostLoop(final AstNode node) {
+    depthNodes.pop();
+  }
+
+  @Override
+  protected void walkPreTry(final AstNode node) {
+    this.checkNestingDepth(node);
+  }
+
+  @Override
+  protected void walkPostTry(final AstNode node) {
+    depthNodes.pop();
+  }
+
+  private void checkNestingDepth(final AstNode node) {
+    depthNodes.push(node);
+
+    if (depthNodes.size() == DEFAULT_MAXIMUM_NESTING_DEPTH + 1) {
+      AstNode lastNode = depthNodes.peek();
+      this.addIssue(lastNode, String.format(MESSAGE, DEFAULT_MAXIMUM_NESTING_DEPTH));
+    }
+  }
```

Comment:
Do we want to add more constructs to the check, like `_block`, `_protect`, `_lock`, `_catch`, or embedded procedures?

Reply:
```magik
_method a.b
	...

	_block
		_if a
		_then
			...
		_endif
	_endblock

	...
_endmethod
```

I think this should be marked, as the depth is increased by the block.

Reply:
And in case of a method, should the top body be counted/

Reply:
I believe you are saying the top should never be counted, whether it is a method or a block. Which is also what I am saying.

Any deeper block (or rather, body) should be counted, I think.

Reply:
I think the easiest is to (only) use the `walkPreBody()` + `walkPostBody()` methods, incrementing and decrementing currentDepth respectively. This way there is no need to keep a stack. But you do keep track of all constructs which increment the nesting level.

Then, ignore the top body. I.e., a method which returns the current depth, something such as:
```java
int getCurrentDepth() {
  return this.currentDepth - 1;
}
```

Reply:
I see, test `testNotExceedingMaximumNestingDepthWithEarlyReturns` fails. This is because the `_if b _then _leave _endif` is also adding a level. Do you expect that this does not add to the nesting depth?

I also see that the check tests for `node.getToken() != null`. If the body is empty, this body is ignored. Following this, several tests have if constructs with empty bodies. Now the nesting depth is already too deep at the lowest level, but perhaps it is better to add a statement to the body.

Reply:
I've added some changes. The question about the test failing still stands!

Reply:
There are (at least) two ways to look at this:
1. It should add to the nesting depth. If you're at that depth, it is either time to refactor the method, or to increase the maximum depth. This is the check in its current (and purest) form.
2. Allow a body with only a `_leave`/`_continue`/`_return` (others?) in these cases.

I think both should be supported, but configurable. Can you add a boolean parameter to the check with which you can configure this?